File: 12 December/12-20/purge_most_frequent.py
from collections import Counter

def purge_most_frequent(arr):
    counts = dict(Counter(arr))
    max_count = 0

    for element in counts:
        if counts[element] > max_count:
            elements_to_remove_set = {element}
            max_count = counts[element]
        elif counts[element] == max_count:
            elements_to_remove_set.add(element)

    return [element for element in arr if element not in elements_to_remove_set]

print(purge_most_frequent([1, 2, 2, 3]))
print(purge_most_frequent(["a", "b", "d", "b", "c", "d", "c", "d", "c", "d"]))
print(purge_most_frequent(["red", "blue", "green", "red", "blue", "green", "blue"]))
print(purge_most_frequent([5, 5, 5, 5]))

print(purge_most_frequent([1, 1, 2, 2, 3]))

# --- TEST SUITE ---

tests_text = '''
1. purge_most_frequent([1, 2, 2, 3]) should return [1, 3].
2. purge_most_frequent(["a", "b", "d", "b", "c", "d", "c", "d", "c", "d"]) should return ["a", "b", "b", "c", "c", "c"].
3. purge_most_frequent(["red", "blue", "green", "red", "blue", "green", "blue"]) should return ["red", "green", "red", "green"].
4. purge_most_frequent([5, 5, 5, 5]) should return [].
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 12 December/12-20/purgeMostFrequent.js
function purgeMostFrequent(arr) {
  const counts = {};
  let maxCount = 0;
  let elementsToRemove = new Set();
  
  arr.forEach(element => counts[element] = (counts[element] || 0) + 1);

  for (let element in counts) {
    if (counts[element] > maxCount) {
      elementsToRemove = new Set().add(parseInt(element) || element);
      maxCount = counts[element];
    }
    else if (counts[element] === maxCount) {
      elementsToRemove.add(parseInt(element) || element);
    }
  }

  return arr.filter(element => !elementsToRemove.has(element));
}

console.log(purgeMostFrequent([1, 2, 2, 3]));
console.log(purgeMostFrequent(["a", "b", "d", "b", "c", "d", "c", "d", "c", "d"]));
console.log(purgeMostFrequent(["red", "blue", "green", "red", "blue", "green", "blue"]));
console.log(purgeMostFrequent([5, 5, 5, 5]));

console.log(purgeMostFrequent([1, 1, 2, 2, 3]));

// --- TEST SUITE ---

const testsText = `
1. purgeMostFrequent([1, 2, 2, 3]) should return [1, 3].
2. purgeMostFrequent(["a", "b", "d", "b", "c", "d", "c", "d", "c", "d"]) should return ["a", "b", "b", "c", "c", "c"].
3. purgeMostFrequent(["red", "blue", "green", "red", "blue", "green", "blue"]) should return ["red", "green", "red", "green"].
4. purgeMostFrequent([5, 5, 5, 5]) should return [].
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);